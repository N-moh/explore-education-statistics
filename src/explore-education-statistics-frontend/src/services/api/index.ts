import axios from 'axios';
import { commaSeparated } from '../util/paramSerializers';

export const baseUrl = {
  content: process.env.CONTENT_API_BASE_URL,
  data: process.env.DATA_API_BASE_URL,
};

export const contentApi = axios.create({
  baseURL: `${baseUrl.content}/api/`,
  paramsSerializer: commaSeparated,
});

export const dataApi = axios.create({
  baseURL: `${baseUrl.data}/api/`,
  paramsSerializer: commaSeparated,
});
