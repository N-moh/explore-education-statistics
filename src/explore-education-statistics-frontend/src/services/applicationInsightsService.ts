import { ApplicationInsights } from '@microsoft/applicationinsights-web';

// eslint-disable-next-line import/prefer-default-export
export const initApplicationInsights = () => {
  if (
    process.env.AI_INSTRUMENTATION !== undefined &&
    process.env.AI_TRACKING === 'true'
  ) {
    const appInsights = new ApplicationInsights({
      config: {
        instrumentationKey: process.env.AI_INSTRUMENTATION,
        enableAutoRouteTracking: true,
        autoTrackPageVisitTime: true,
      },
    });
    appInsights.loadAppInsights();

    // eslint-disable-next-line no-console
    console.log('Application Insights initialised');
  }
};
