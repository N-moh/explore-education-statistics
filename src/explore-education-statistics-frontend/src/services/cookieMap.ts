function daysFromNow(days = 0) {
  const result = new Date();
  if (!days) {
    // Add one month
    result.setMonth([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 0][result.getMonth()]);
  } else {
    result.setDate(result.getDate() + days);
  }
  return result;
}

const CookieMap = {
  bannerSeenCookie: {
    name: 'ees_banner_seen',
    expires: daysFromNow(),
    duration: '1 month',
  },
  disableGACookie: {
    name: 'ees_disable_google_analytics',
    expires: daysFromNow(),
    duration: '1 month',
  },
};

export default CookieMap;
