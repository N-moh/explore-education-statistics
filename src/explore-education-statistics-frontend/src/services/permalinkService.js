"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var api_1 = require("@common/services/api");
exports.default = {
    getPermalink: function (publicationSlug) {
        return api_1.dataApi.get("Permalink/" + publicationSlug);
    },
    createTablePermalink: function (query) {
        return api_1.dataApi.post('/permalink', query);
    },
};
