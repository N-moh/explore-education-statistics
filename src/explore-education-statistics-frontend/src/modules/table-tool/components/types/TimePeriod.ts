import { TimePeriodOption } from '@common/services/tableBuilderService';
import { TimePeriodCode } from '@common/services/types/TimePeriod';

export default class TimePeriod {
  public readonly year: number;

  public readonly code: string;

  public readonly label: string;

  public constructor({ year, code, label }: TimePeriodOption) {
    this.code = code;
    this.year = year;
    this.label = label;
  }

  public get value(): string {
    return `${this.year}_${this.code}`;
  }
}

export function parseYearCodeTuple(value: string): [number, string] {
  const [year, code] = value.split('_');

  const parsedYear = Number(year);
  const parsedCode = code as TimePeriodCode;

  if (Number.isNaN(parsedYear)) {
    throw new TypeError('Could not parse time period year');
  }

  return [parsedYear, parsedCode];
}
