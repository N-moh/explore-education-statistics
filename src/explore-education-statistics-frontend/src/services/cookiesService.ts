import * as googleAnalyticsService from '@frontend/services/googleAnalyticsService';

interface CookieObj {
  name: string;
  value?: any;
  expires?: Date;
}

function oneMonthFromNow() {
  const dateToday = new Date();
  return new Date(dateToday.setMonth(dateToday.getMonth() + 1));
}

export const bannerSeenCookieName = 'ees_banner_seen';
export const cookieSettingsCookieName = 'ees_cookie_settings';

export function getCookies() {
  // turns document.cookie e.g. 'cookie_name=hello-world; cookietwo=true'
  // into { cookie_name: "hellow-world", cookietwo: true }
  const cookies: any = {};
  document.cookie.split('; ').forEach(cookiesPairString => {
    const [key, value]: string[] = cookiesPairString.split('=');
    cookies[key] = value;
  });
  return cookies;
}

export function setCookies(newCookies: CookieObj[]) {
  newCookies.forEach(cookie => {
    const { name, value, expires = oneMonthFromNow() } = cookie;
    document.cookie = `${name}=${JSON.stringify(value)}; expires=${expires}`;

    if (name === cookieSettingsCookieName) {
      const cookieSettingGA = value.googleAnalytics;
      if (cookieSettingGA && cookieSettingGA.length) {
        if (cookieSettingGA === 'off') {
          googleAnalyticsService.disableGA();
        } else {
          googleAnalyticsService.enableGA();
        }
      }
    }
  });
}

export function deleteCookies(cookieNames: string[]) {
  setCookies(
    cookieNames.map(cookieName => {
      return {
        name: cookieName,
        expires: new Date(0),
      };
    }),
  );
}

export function acceptCookies() {
  // Accepted cookies banner
  setCookies([
    {
      name: bannerSeenCookieName,
      value: 'true',
    },
  ]);
}
