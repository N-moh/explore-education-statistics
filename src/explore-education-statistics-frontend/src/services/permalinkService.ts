import { dataApi } from '@common/services/api';
import FullTable from '@frontend/modules/table-tool/components/types/FullTable';

export interface Permalink {
  id: string;
  title: string;
  created: string;
  fullTable: FullTable;
}

export default {
  getPermalink(publicationSlug: string): Promise<Permalink> {
    return dataApi.get(`Permalink/${publicationSlug}`);
  },
};
