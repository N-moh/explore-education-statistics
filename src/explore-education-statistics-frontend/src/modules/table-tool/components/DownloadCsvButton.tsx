import ButtonText from '@common/components/ButtonText';
import cartesian from '@common/lib/utils/cartesian';
import {
  PublicationSubjectMeta,
  TableData,
} from '@common/services/tableBuilderService';
import { Dictionary } from '@common/types';
import {
  CategoryFilter,
  Indicator,
  LocationFilter,
  Filter,
} from '@frontend/modules/table-tool/components/types/filters';
import TimePeriod from '@frontend/modules/table-tool/components/types/TimePeriod';
import { saveAs } from 'file-saver';
import Papa from 'papaparse';
import React from 'react';

interface Props {
  publicationSlug: string;
  meta: PublicationSubjectMeta;
  indicators: Indicator[];
  filters: Dictionary<CategoryFilter[]>;
  timePeriods: TimePeriod[];
  locations: LocationFilter[];
  results: TableData['results'];
}

const DownloadCsvButton = ({
  publicationSlug,
  meta,
  indicators,
  filters,
  timePeriods,
  locations,
  results,
}: Props) => {
  const getCsvData = (): string[][] => {
    const filterColumns = Object.entries(filters).map(
      ([key]) => meta.filters[key].legend,
    );

    const indicatorColumns = indicators.map(indicator => {
      const unit = indicator.unit ? ` (${indicator.unit})` : '';
      return `${indicator.label}${unit}`;
    });

    const columns = [
      'Location',
      'Time period',
      ...filterColumns,
      ...indicatorColumns,
    ];

    const rows = cartesian<Filter>(
      locations,
      timePeriods,
      ...Object.values(filters),
    ).map(row => {
      // TODO: Remove ignore when Prettier stops adding trailing comma to tuple type
      // prettier-ignore
      const [location, timePeriod, ...filterOptions] = row as [
        LocationFilter,
        TimePeriod,
        ...CategoryFilter[]
      ];

      const indicatorCells = indicators.map(indicator => {
        const matchingResult = results.find(result => {
          return Boolean(
            filterOptions.every(filter =>
              result.filters.includes(filter.value),
            ) &&
              result.timePeriod === timePeriod.value &&
              result.location[location.level] &&
              result.location[location.level].code === location.value,
          );
        });

        if (!matchingResult) {
          return 'n/a';
        }

        return matchingResult.measures[indicator.value];
      });

      return [...row.map(column => column.label), ...indicatorCells];
    });

    return [columns, ...rows];
  };

  return (
    <ButtonText
      onClick={() => {
        const data = new Blob([Papa.unparse(getCsvData())], {
          type: 'text/csv;charset=utf-8',
        });

        const filename = `data-${publicationSlug}.csv`;

        saveAs(data, filename, {
          autoBom: true,
        });
      }}
    >
      Download underlying data (.csv)
    </ButtonText>
  );
};

export default DownloadCsvButton;
