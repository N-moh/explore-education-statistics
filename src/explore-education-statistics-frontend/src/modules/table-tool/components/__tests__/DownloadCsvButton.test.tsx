import Papa from 'papaparse';
import React from 'react';
import { fireEvent, render } from 'react-testing-library';
import TimePeriod from '@common/services/types/TimePeriod';
import { FullTable } from '@frontend/services/permalinkService';
import DownloadCsvButton from '../DownloadCsvButton';

describe('DownloadCsvButton', () => {
  const emptyTable: FullTable = {
    subjectMeta: {
      publicationName: '',
      subjectId: '',
      subjectName: '',
      footnotes: [],
      filters: {
        characteristics: {
          legend: 'Characteristics',
          hint: '',
          options: {
            gender: {
              label: 'Gender',
              options: [
                { label: 'Male', value: 'gender_male' },
                { label: 'Female', value: 'gender_female' },
              ],
            },
          },
        },
        schoolType: {
          legend: 'School type',
          hint: '',
          options: {
            default: {
              label: '',
              options: [
                {
                  label: 'State-funded-primary',
                  value: 'school_primary',
                },
                {
                  label: 'State-funded secondary',
                  value: 'school_secondary',
                },
              ],
            },
          },
        },
      },
      indicators: [
        {
          label: 'Authorised absence rate',
          value: 'auth_abs_rate',
          unit: '%',
        },
        {
          label: 'Number of authorised absence sessions',
          value: 'auth_abs_sess',
          unit: '',
        },
      ],
      locations: [
        {
          value: 'england',
          label: 'England',
          level: 'country',
        },
        { value: 'south_yorkshire', label: 'South Yorkshire', level: 'region' },
      ],
      timePeriodRange: [
        TimePeriod.fromString('2014_AY'),
        TimePeriod.fromString('2015_AY'),
      ],
    },
    results: [],
  };

  test('clicking generates complete csv', async () => {
    const { getByText } = render(
      <DownloadCsvButton
        publicationSlug="pupil-absence"
        fullTable={{
          ...emptyTable,
          results: [
            {
              filters: ['gender_male', 'school_primary'],
              timePeriod: '2014_AY',
              location: {
                country: {
                  code: 'england',
                  name: 'England',
                },
              },
              measures: {
                authAbsRate: '1.2',
                authAbsSess: '2',
              },
            },
            {
              filters: ['gender_male', 'school_secondary'],
              timePeriod: '2014_AY',
              location: {
                country: {
                  code: 'england',
                  name: 'England',
                },
              },
              measures: {
                authAbsRate: '3.4',
                authAbsSess: '4',
              },
            },
            {
              filters: ['gender_female', 'school_primary'],
              timePeriod: '2014_AY',
              location: {
                country: {
                  code: 'england',
                  name: 'England',
                },
              },
              measures: {
                authAbsRate: '5.6',
                authAbsSess: '6',
              },
            },
            {
              filters: ['gender_female', 'school_secondary'],
              timePeriod: '2014_AY',
              location: {
                country: {
                  code: 'england',
                  name: 'England',
                },
              },
              measures: {
                authAbsRate: '7.8',
                authAbsSess: '8',
              },
            },
            {
              filters: ['gender_male', 'school_primary'],
              timePeriod: '2015_AY',
              location: {
                country: {
                  code: 'england',
                  name: 'England',
                },
              },
              measures: {
                authAbsRate: '9',
                authAbsSess: '10',
              },
            },
            {
              filters: ['gender_male', 'school_secondary'],
              timePeriod: '2015_AY',
              location: {
                country: {
                  code: 'england',
                  name: 'England',
                },
              },
              measures: {
                authAbsRate: '11.2',
                authAbsSess: '12',
              },
            },
            {
              filters: ['gender_female', 'school_primary'],
              timePeriod: '2015_AY',
              location: {
                country: {
                  code: 'england',
                  name: 'England',
                },
              },
              measures: {
                authAbsRate: '13.4',
                authAbsSess: '14',
              },
            },
            {
              filters: ['gender_female', 'school_secondary'],
              timePeriod: '2015_AY',
              location: {
                country: {
                  code: 'england',
                  name: 'England',
                },
              },
              measures: {
                authAbsRate: '15.6',
                authAbsSess: '16',
              },
            },
            {
              filters: ['gender_male', 'school_primary'],
              timePeriod: '2014_AY',
              location: {
                region: {
                  code: 'south_yorkshire',
                  name: 'South Yorkshire',
                },
              },
              measures: {
                authAbsRate: '16.7',
                authAbsSess: '17',
              },
            },
            {
              filters: ['gender_male', 'school_secondary'],
              timePeriod: '2014_AY',
              location: {
                country: {
                  code: 'england',
                  name: 'England',
                },
              },
              measures: {
                authAbsRate: '17.8',
                authAbsSess: '18',
              },
            },
            {
              filters: ['gender_female', 'school_primary'],
              timePeriod: '2014_AY',
              location: {
                country: {
                  code: 'england',
                  name: 'England',
                },
              },
              measures: {
                authAbsRate: '18.9',
                authAbsSess: '19',
              },
            },
            {
              filters: ['gender_female', 'school_secondary'],
              timePeriod: '2014_AY',
              location: {
                country: {
                  code: 'england',
                  name: 'England',
                },
              },
              measures: {
                authAbsRate: '20.1',
                authAbsSess: '20',
              },
            },
            {
              filters: ['gender_male', 'school_primary'],
              timePeriod: '2014_AY',
              location: {
                country: {
                  code: 'england',
                  name: 'England',
                },
              },
              measures: {
                authAbsRate: '22.3',
                authAbsSess: '22',
              },
            },
            {
              filters: ['gender_male', 'school_secondary'],
              timePeriod: '2014_AY',
              location: {
                country: {
                  code: 'england',
                  name: 'England',
                },
              },
              measures: {
                authAbsRate: '24.5',
                authAbsSess: '25',
              },
            },
            {
              filters: ['gender_female', 'school_primary'],
              timePeriod: '2014_AY',
              location: {
                country: {
                  code: 'england',
                  name: 'England',
                },
              },
              measures: {
                authAbsRate: '26.7',
                authAbsSess: '27',
              },
            },
            {
              filters: ['gender_female', 'school_secondary'],
              timePeriod: '2014_AY',
              location: {
                country: {
                  code: 'england',
                  name: 'England',
                },
              },
              measures: {
                authAbsRate: '28.9',
                authAbsSess: '29',
              },
            },
          ],
        }}
      />,
    );

    jest.spyOn(Papa, 'unparse');

    fireEvent.click(getByText('Download underlying data (.csv)'));

    const csv = (Papa.unparse as jest.Mock).mock.calls[0][0];

    expect(csv).toMatchSnapshot();
  });

  test('generated csv does not format values', () => {
    const { getByText } = render(
      <DownloadCsvButton
        publicationSlug="pupil-absence"
        fullTable={{
          ...emptyTable,
          results: [
            {
              filters: ['gender_female', 'school_primary'],
              timePeriod: '2015_AY',
              location: {
                country: {
                  code: 'england',
                  name: 'England',
                },
              },
              measures: {
                authAbsRate: '12300000',
                authAbsSess: '44255667.2356',
              },
            },
          ],
        }}
      />,
    );

    jest.spyOn(Papa, 'unparse');

    fireEvent.click(getByText('Download underlying data (.csv)'));

    const csv = (Papa.unparse as jest.Mock).mock.calls[0][0];

    expect(csv).toMatchSnapshot();
  });

  test('generated csv can contain suppressed or cells with no result ', () => {
    const { getByText } = render(
      <DownloadCsvButton
        publicationSlug="pupil-absence"
        fullTable={{
          ...emptyTable,
          results: [
            {
              filters: ['gender_female', 'school_primary'],
              timePeriod: '2015_AY',
              location: {
                country: {
                  code: 'england',
                  name: 'England',
                },
              },
              measures: {
                authAbsRate: '13.4',
                authAbsSess: 'x',
              },
            },
          ],
        }}
      />,
    );

    jest.spyOn(Papa, 'unparse');

    fireEvent.click(getByText('Download underlying data (.csv)'));

    const csv = (Papa.unparse as jest.Mock).mock.calls[0][0];

    expect(csv).toMatchSnapshot();
  });
});
