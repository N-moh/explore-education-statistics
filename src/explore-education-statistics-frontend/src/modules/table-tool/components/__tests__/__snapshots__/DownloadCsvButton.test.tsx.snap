// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DownloadCsvButton clicking generates complete csv 1`] = `
Array [
  Array [
    "Time period",
    "Characteristics",
    "School type",
    "Authorised absence rate (%)",
    "Number of authorised absence sessions",
  ],
  Array [
    "2014/15",
    "Male",
    "State-funded primary",
    "1.2",
    "2",
  ],
  Array [
    "2014/15",
    "Male",
    "State-funded secondary",
    "3.4",
    "4",
  ],
  Array [
    "2014/15",
    "Female",
    "State-funded primary",
    "5.6",
    "6",
  ],
  Array [
    "2014/15",
    "Female",
    "State-funded secondary",
    "7.8",
    "8",
  ],
  Array [
    "2015/16",
    "Male",
    "State-funded primary",
    "9",
    "10",
  ],
  Array [
    "2015/16",
    "Male",
    "State-funded secondary",
    "11.2",
    "12",
  ],
  Array [
    "2015/16",
    "Female",
    "State-funded primary",
    "13.4",
    "14",
  ],
  Array [
    "2015/16",
    "Female",
    "State-funded secondary",
    "15.6",
    "16",
  ],
]
`;

exports[`DownloadCsvButton generated csv can contain suppressed or cells with no result  1`] = `
Array [
  Array [
    "Time period",
    "Characteristics",
    "School type",
    "Authorised absence rate (%)",
    "Number of authorised absence sessions",
  ],
  Array [
    "2015/16",
    "Male",
    "State-funded primary",
    "n/a",
    "n/a",
  ],
  Array [
    "2015/16",
    "Female",
    "State-funded primary",
    "13.4",
    "x",
  ],
]
`;

exports[`DownloadCsvButton generated csv does not format values 1`] = `
Array [
  Array [
    "Time period",
    "Characteristics",
    "School type",
    "Authorised absence rate (%)",
    "Number of authorised absence sessions",
  ],
  Array [
    "2015/16",
    "Female",
    "State-funded primary",
    "12300000",
    "44255667.2356",
  ],
]
`;
